bundles/admin/admin-layereditor/instance.js

```diff
@@ -341,16 +364,24 @@ Oskari.clazz.defineES('Oskari.admin.admin-layereditor.instance',
             });
             this.themeFlyout.setSaveAction((value, id) => {
                 const httpMethod = id ? 'POST' : 'PUT';
-                const payload = id ? { locales: value, id: id } : { locales: value };
+                const getpayload = () => {
+                    if (id) {
+                        return parentId ? { locales: value, id: id, parentId: parentId } : { locales: value, id: id };
+                    } else {
+                        return { locales: value };
+                    }
```

Comment:
Saving a new group (group without id) will never send the parent id? So you can only add groups to root? But then you have another saveAction for subgroups? What does parent id mean here? Why do group and subgroup need different handling? Aren't they just groups that either have a parent or don't?

Reply:
To my understanding payload needs to have root id (id) and parent id (parentId) when saving "subsubgroups".
Parent id points to id at one step backwards on hierarchy.
This might be wrong and parentId here is unnecessary.

Reply:
Sure, looks like it needs parentId for update and insert for subgroups. I'm wondering why it's never sent for new groups (groups without id).

Reply:
Also there's another setSaveAction() that uses parent id differently so can they be combined?

Reply:
I feel like there's some hundred lines of copy pasted code here that isn't really needed and makes maintaining this stuff harder than it should be.

Reply:
It has been a bit hard to overcome the fact that so much functionality has been packed to getThemeFlyout method so additional getSubthemeFlyout method was created (copy pasted and modified). For now this has been fixed and previously added code has been removed and only getThemeFlyout is used now with some additional props letting code know if user has pressed "Add theme", "Modify Theme" or "Add subtheme". Generally I feel like this would be so much simpler if all of these three methods would be separately done when this method was implemented at first place.

Reply:
Sure, though in my opinion adding is basically the same thing as modifying but you either have an id for modification or you don't hence you are adding. In the same sense subtheme doesn't really have to be any different than a "root" theme. A theme either has a parent id (making it a "subtheme") or it doesn't  (making it a "root" theme) and that can be also used to determine if you want to show a different title for the form or not. The theme/group details are loaded from the same endpoint whether its a root or subgroup and clicking save or delete are also handled by the same endpoint regardless. I'm not sure that having that separation adds any value for the end-user either but maybe it does.

This is an open source project so if you feel too much code is packed to a function you have the ability to make it better. That is also what I'm trying to do here. Instead of blindly accepting and merging PRs I try to guide the code towards a better future. Of course I'm not always right but it is my goal to make the code better and more maintainable since I'm one of the devs that actually needs to maintain it. If you have multiple places doing essentially the same thing it's very easy to miss some of those places when a modification is required. That leads to bugs that are not easily spotted. If code needs to be copy-pasted in large amounts it needs some comments on the code to document WHY that code was copy-pasted. Otherwise when you need to do something with that code again spotting that _one line difference_ in a 100 lines of a function that has been copy-pasted to four different functions is really painful. It usually leads to the function being copied again and again for fear of breaking something making it worse and worse for the future. I agree that it's much easier to keep adding code than making sure that a function does what it is supposed to do based with different parameters. However it's much easier to maintain code where you only need to handle a change on one spot than copy-pasting that same change in four more spots.

In this example only things that change between adding/editing, root/subtheme is that the HTTP-method changes and the id and parent id are either known and sent or unknown and as such can't be sent. Copy pasting some 100 lines of code to a parallel functions just for those changes feels wrong to me:

- If we have id -> use HTTP POST, otherwise use PUT
- If we have an id -> pass it to the backend
- If we have a parent id -> pass it to the backend

You could even argue that you don't need a control-flow for the last two and just send them to the backend as received in the payload and as there's no value the backend will handle accordingly whether you send them or not.